src/main/java/io/quarkus/gizmo2/impl/AnnotationCreatorImpl.java

```diff
@@ -0,0 +1,528 @@
+    }
+
+    @Override
+    public void with(String name, char value) {
+        with(name, AnnotationValue.ofChar(value));
+    }
+
+    @Override
+    public void with(CharProperty<A> prop, char value) {
+        with(Util.serializedLambda(prop), AnnotationValue.ofChar(value));
+    }
+
+    @Override
+    public void with(String name, String value) {
+        with(name, AnnotationValue.ofString(value));
+    }
+
+    @Override
+    public void with(StringProperty<A> prop, String value) {
+        with(Util.serializedLambda(prop), AnnotationValue.ofString(value));
+    }
+
+    @Override
+    public void with(String name, Class<?> value) {
+        with(name, AnnotationValue.ofClass(Util.classDesc(value)));
+    }
+
+    @Override
+    public void with(ClassProperty<A> prop, Class<?> value) {
+        with(Util.serializedLambda(prop), AnnotationValue.ofClass(Util.classDesc(value)));
+    }
+
+    @Override
+    public void with(String name, ClassDesc value) {
+        with(name, AnnotationValue.ofClass(value));
+    }
+
+    @Override
+    public void with(ClassProperty<A> prop, ClassDesc value) {
+        with(Util.serializedLambda(prop), AnnotationValue.ofClass(value));
+    }
+
+    @Override
+    public <E extends Enum<E>> void with(String name, E value) {
+        with(name, AnnotationValue.ofEnum(Util.classDesc(value.getDeclaringClass()), value.name()));
+    }
+
+    @Override
+    public <E extends Enum<E>> void with(EnumProperty<A, E> prop, E value) {
+        with(Util.serializedLambda(prop), AnnotationValue.ofEnum(Util.classDesc(value.getDeclaringClass()), value.name()));
+    }
+
+    @Override
+    public void with(String name, ClassDesc enumClass, String enumConstant) {
+        with(name, AnnotationValue.ofEnum(enumClass, enumConstant));
+    }
+
+    @Override
+    public <S extends Annotation> void with(String name, Class<S> annotationClass, Consumer<AnnotationCreator<S>> builder) {
+        with(name, AnnotationValue.ofAnnotation(makeAnnotation(annotationClass, builder)));
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <S extends Annotation> void with(AnnotationProperty<A, S> prop, Consumer<AnnotationCreator<S>> builder) {
+        // the impl method is always an annotation element, which never takes any parameter,
+        // so the signature starts with `()` and the rest is the descriptor of the desired annotation class
+        String sig = Util.serializedLambda(prop).getImplMethodSignature();
+        assert sig.startsWith("()L") && sig.endsWith(";");
+        String annotationClass = sig.substring(3, sig.length() - 1).replace('/', '.');
+        Class<S> clazz;
+        try {
+            ClassLoader cl = Thread.currentThread().getContextClassLoader();
+            if (cl == null) {
+                cl = AnnotationCreator.class.getClassLoader();
+            }
+            clazz = (Class<S>) Class.forName(annotationClass, false, cl);
```

Comment:
I think we should consider requiring that a `Class<S>` be passed in instead of trying to do class loader magic. TCCL is almost certainly the wrong one to use at any rate.

Reply:
Ideally yes, but in practice I can't think of a totally safe/totally correct way to do it automatically. The lambda instance does not seem to have a concrete generic type, so we can't probe it that way. AFAICT we can't know for sure what class loader originated the type of the nested annotation.

Reply:
Actually, I'm being silly. We can examine the method on the enclosing annotation type.

Reply:
Done.